engines/e_pkcs11.c

```diff
@@ -0,0 +1,808 @@
+            PKCS11_trace("C_Login failed, error: %#08X\n", rv);
+            PKCS11err(PKCS11_F_PKCS11_LOGIN, PKCS11_R_LOGIN_FAILED);
+            goto err;
+        }
+    return 1;
+    }
+ err:
+    return 0;
+}
+
+static int pkcs11_logout(CK_SESSION_HANDLE session)
+{
+    CK_RV rv;
+
+    rv = pkcs11_funcs->C_Logout(session);
+    if (rv != CKR_USER_NOT_LOGGED_IN && rv != CKR_OK) {
+        PKCS11_trace("C_Logout failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_LOGOUT, PKCS11_R_LOGOUT_FAILED);
+        goto err;
+    }
+    return 1;
+
+ err:
+    return 0;
+}
+
+static void pkcs11_end_session(CK_SESSION_HANDLE session)
+{
+    pkcs11_funcs->C_CloseSession(session);
+}
+
+CK_OBJECT_HANDLE pkcs11_get_private_key(PKCS11_CTX *ctx)
+{
+    CK_RV rv;
+    CK_OBJECT_CLASS key_class = CKO_PRIVATE_KEY;
+    CK_KEY_TYPE key_type = CKK_RSA;
+    CK_OBJECT_HANDLE objhandle;
+    unsigned long count;
+    CK_ATTRIBUTE tmpl[3];
+
+    tmpl[0].type = CKA_CLASS;
+    tmpl[0].pValue = &key_class;
+    tmpl[0].ulValueLen = sizeof(key_class);
+    tmpl[1].type = CKA_KEY_TYPE;
+    tmpl[1].pValue = &key_type;
+    tmpl[1].ulValueLen = sizeof(key_type);
+    if (ctx->id != NULL) {
+        tmpl[2].type = CKA_ID;
+        tmpl[2].pValue = ctx->id;
+        tmpl[2].ulValueLen = strlen((char *)ctx->id);
+    } else {
+        tmpl[2].type = CKA_LABEL;
+        tmpl[2].pValue = ctx->label;
+        tmpl[2].ulValueLen = strlen((char *)ctx->label);
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsInit(ctx->session, tmpl, OSSL_NELEM(tmpl) );
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsInit failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_INIT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjects(ctx->session, &objhandle, 1, &count);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjects failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsFinal(ctx->session);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsFinal failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FINAL_FAILED);
+        goto err;
+    }
+
+    return objhandle;
+
+ err:
+    return 0;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    size_t vlen, i, j = 0, tmplen;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    tmplen = strlen(tmp);
+    *value = OPENSSL_malloc(tmplen + 1);
+
+    if (*value == NULL) goto err;
```

Comment:
Please insert a linebreak before the `goto err;` statement. Having it on the same line makes it impossible to set a debugger breakpoint on the error case. Also, with only one exception, it is always placed on a separate line in the OpenSSL source code.

This holds for several locations.